File: project/src/main/java/javaspark/joinlianxi.java
//import org.apache.spark.SparkConf;
//import org.apache.spark.api.java.JavaPairRDD;
//import org.apache.spark.api.java.JavaSparkContext;
//import org.apache.spark.api.java.function.VoidFunction;
//import scala.Tuple2;
//import java.util.List;
//import java.util.Arrays;
//public class joinlianxi {
//    public static void main(String[] args) {
//
//    }
//    private static void join(){
//        SparkConf conf=new SparkConf()
//                .setAppName("join")
//                .setMaster("local");
//        JavaSparkContext sc=new JavaSparkContext(conf);
//        List<Tuple2<Integer,String>> scoreList1= Arrays.asList(
//                new Tuple2<Integer,String>(1,"tom"),
//                new Tuple2<Integer,String>(2,"momo"),
//                new Tuple2<Integer,String>(3,"tentent"),
//                new Tuple2<Integer,String>(4,"didi")
//        );
//        List<Tuple2<Integer,Integer>> scoreList2= Arrays.asList(
//                new Tuple2<Integer,Integer>(1,10),
//                new Tuple2<Integer,Integer>(2,20),
//                new Tuple2<Integer,Integer>(3,30),
//                new Tuple2<Integer,Integer>(4,40)
//        );
//        JavaPairRDD<Integer,String> scorelistrdd1=sc.parallelizePairs(scoreList1);
//        JavaPairRDD<Integer,String> scorelistrdd2=sc.parallelizePairs(scoreList1);
//
//        JavaPairRDD<Integer,Tuple2<Integer,String>>studentScores=scoreList1.join(scoreList2);
//        studentScores.foreach(
//                new VoidFunction <Tuple2 <Integer, Tuple2 <Integer, String>>>() {
//                    @Override
//                    public void call(Tuple2 <Integer, Tuple2 <Integer, String>> t) throws Exception {
//                        System.out.println("student id:"+t._1);
//                        System.out.println("student name:"+t._2._1);
//                        System.out.println("student score:"+t._2._2);
//                        System.out.printf("=======================");
//                    }
//                }
//
//
//        );
//        sc.close();
//
//
//
//
//
//
//
//
//
//
//    }
//
//
//
//
//
//
//
//
//}

File: project/src/main/scala/day20190319/joinlianxi.scala
package day20190319

import org.apache.spark.{SparkConf, SparkContext}

object joinlianxi {
  def main(args: Array[String]): Unit = {
    //join()
    cogroup()
  }
  def join(){
    val conf = new SparkConf()
      .setAppName("join")
      .setMaster("local")
    val sc =new  SparkContext(conf)
    val studentList =Array(Tuple2(1, "tom"), Tuple2(2, "jerry"), Tuple2(3, "marry"))
    val scoreList =Array(Tuple2(1, 12), Tuple2(2, 43), Tuple2(3, 54))
    val students = sc.parallelize(studentList)
    val scores = sc.parallelize(scoreList)
    val studentscores = students.join(scores)
    studentscores.foreach(studentscore =>{
      println("学生id：" + studentscore._1)
      println("学生名字：" + studentscore._2._1)
      println("学生分数：" + studentscore._2._2)})
      println("==================================")
  }
  def cogroup(){
    val conf = new SparkConf()
      .setAppName("cogroup")
      .setMaster("local")
    val sc =new  SparkContext(conf)
    val studentList =Array(Tuple2(1, "tom"), Tuple2(1, "jerry"), Tuple2(3, "marry"))
    val scoreList =Array(Tuple2(1, 12), Tuple2(2, 43), Tuple2(3, 54),Tuple2(1, 34),Tuple2(1, 54))
    val students = sc.parallelize(studentList)
    val count=students.countByKey()
    println(count)
    count.map(x=>println(x))
    val scores = sc.parallelize(scoreList)


    val studentscores = students.cogroup(scores)
    studentscores.foreach(studentscore =>{
      println("学生id：" + studentscore._1)
      println("学生名字：" + studentscore._2._1)
      println("学生分数：" + studentscore._2._2)})
    println("==================================")
  }
}
